stac_mjx/controller.py

```diff
@@ -1,330 +1,546 @@
+        self._trunk_kps = jp.array(
+            [n in self.model_cfg["TRUNK_OPTIMIZATION_KEYPOINTS"] for n in kp_names],
         )
 
-    utils.params["indiv_parts"] = indiv_parts
-
-
-def create_keypoint_sites(root):
-    """Create sites for keypoints (used for rendering).
-
-    Args:
-        root (mjcf.Element): root element of mjcf
-
-    Returns:
-        (dmcontrol.Physics, mujoco.Model, [mjcf.Element]): physics, mjmodel, and list of the created sites
-    """
-    keypoint_sites = []
-    # set up keypoint rendering by adding the kp sites to the root body
-    for id, name in enumerate(utils.params["KEYPOINT_MODEL_PAIRS"]):
-        start = (np.random.rand(3) - 0.5) * 0.001
-        rgba = utils.params["KEYPOINT_COLOR_PAIRS"][name]
-        site = root.worldbody.add(
-            "site",
-            name=name + "_kp",
-            type="sphere",
-            size=[0.005],
-            rgba=rgba,
-            pos=start,
-            group=2,
+        mj_model.opt.solver = {
+            "cg": mujoco.mjtSolver.mjSOL_CG,
+            "newton": mujoco.mjtSolver.mjSOL_NEWTON,
+        }[stac_cfg.mujoco.solver.lower()]
+
+        mj_model.opt.iterations = stac_cfg.mujoco.iterations
+        mj_model.opt.ls_iterations = stac_cfg.mujoco.ls_iterations
+
+        # Runs faster on GPU with this
+        mj_model.opt.jacobian = 0  # dense
+
+        self._mj_model = mj_model
+
+        # Set joint bounds
+        self._lb = jp.minimum(
+            jp.concatenate([-jp.inf * jp.ones(7), self._mj_model.jnt_range[1:][:, 0]]),
+            0.0,
         )
-        keypoint_sites.append(site)
-
-    physics = mjcf.Physics.from_mjcf_model(root)
-
-    # return physics, mj_model, and sites (to use in bind())
-    return physics, physics.model.ptr, keypoint_sites
-
-
-def set_keypoint_sites(physics, sites, kps):
-    """Bind keypoint sites to physics model.
-
-    Args:
-        physics (_type_): dmcontrol physics object
-        sites (_type_): _description_
-        kps (_type_): _description_
-
-    Returns:
-        (dmcontrol.Physics, mujoco.Model): update physics and model with update site pos
-    """
-    physics.bind(sites).pos[:] = np.reshape(kps.T, (-1, 3))
-    return physics, physics.model.ptr
-
-
-def create_body_sites(root: mjcf.Element):
-    """Create body site elements using dmcontrol mjcf for each keypoint.
-
-    Args:
-        root (mjcf.Element):
-
-    Returns:
-        dmcontrol.Physics, mujoco.Model:
-    """
-    body_sites = []
-    for key, v in utils.params["KEYPOINT_MODEL_PAIRS"].items():
-        parent = root.find("body", v)
-        pos = utils.params["KEYPOINT_INITIAL_OFFSETS"][key]
-        site = parent.add(
-            "site",
-            name=key,
-            type="sphere",
-            size=[0.005],
-            rgba="0 0 0 1",
-            pos=pos,
-            group=3,
+        self._ub = jp.concatenate(
+            [jp.inf * jp.ones(7), self._mj_model.jnt_range[1:][:, 1]]
```

Comment:
First 7 values for the `qpos` bounds are (-inf, inf) because they represent the root position (3) + quaternion (4). How should this constant be managed? Could have constants at the top of the file?
```python
_ROOT_QPOS_LB = -jp.inf * jp.ones(7)
_ROOT_QPOS_UB = jp.inf * jp.ones(7)
```